content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,237 @@
+
+In this example though, it's more important to update the text box with the user input as opposed to updating the charts. A version that prioritizes the input will appear a lot more responsive to the end user although the same computation time is required:
+
+<video src="/blog/scheduling-in-react/concurrent-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+Unfortunately, current user interface architectures makes it non trivial to implement this kind of prioritization. One way to solve this is by [debouncing](https://davidwalsh.name/javascript-debounce-function) the chart update. The problem with this is that the charts still render synchronously when the debounced callback fires, which will again cause our user interface to take multiple seconds during which it is not responsive. We can do better!
+
+## Browser Event Loop
+
+Before we learn more about how this can be achieved, let's dig deeper and understand why the browser has issues with these kind of user interfaces.
+
+JavaScript code is executed in one thread – meaning that only one line of JavaScript can be run at any given time. The same thread is also responsible for other document lifecycles like layout and paint[^1]. This means that whenever JavaScript code runs, the browser is blocked form doing anything else.
+
+To keep the user interface responsive, we only have a very short time frame before we need to be able to receive the next input events. In the browser run loop visualization presented in Subhie Panicker’s and Jason Miller’s talk [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread) at the Chrome Dev Summit 2018 below, we can see that we only have 16 milliseconds (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:
+
+![The browser event loop starts by running input handlers, followed by animation frame callbacks, and ends with document lifecycles (style, layout, paint). All of this should complete within one frame which is approximately 16 milliseconds on a 60Hz display.](event-loop-browser.png)
+
+Most JavaScript frameworks (including the current version of React) will run updates synchronously. We can think of this as a function `render()` which will only return once the DOM was updated. During this time, the main thread is blocked.
+
+## Problems With Current Solutions
+
+With the information above, we can formulate two problems that we have to solve in to get to more responsive user interfaces:
+
+1. **Long running tasks cause frame drops.** We need to make sure that all of our tasks are small and can be completed within a couple of milliseconds so that we can run them within one frame.
+
+2. **Different tasks have different importance.** In the example application above we saw that prioritizing the user input leads to a better experience overall. To do this, we need a way to schedule work.
```

Comment:
importance vs priorities -> not sure what is the better word here, since you use "priority" everywhere else then